src/wallet/db.cpp

```diff
@@ -349,13 +349,26 @@ void BerkeleyEnvironment::CheckpointLSN(const std::string& strFile)
 
 BerkeleyBatch::BerkeleyBatch(BerkeleyDatabase& database, const char* pszMode, bool fFlushOnCloseIn) : pdb(nullptr), activeTxn(nullptr)
 {
+    database.Open(pszMode);
     fReadOnly = (!strchr(pszMode, '+') && !strchr(pszMode, 'w'));
     fFlushOnClose = fFlushOnCloseIn;
     env = database.env.get();
-    if (database.IsDummy()) {
+    pdb = database.m_db.get();
+    strFile = database.strFile;
+    if (strchr(pszMode, 'c') != nullptr && !Exists(std::string("version"))) {
+        bool fTmp = fReadOnly;
+        fReadOnly = false;
+        Write(std::string("version"), CLIENT_VERSION);
+        fReadOnly = fTmp;
+    }
+}
+
+void BerkeleyDatabase::Open(const char* pszMode)
+{
+    m_read_only = (!strchr(pszMode, '+') && !strchr(pszMode, 'w'));
```

Comment:
In 3cdb7fd9d1e92f07c04c325c9b585ff736162198: it's a bit odd that you're introducing `m_read_only` here because it's not read anywhere yet. 64a5a29d692378df443ae599b81c1f4547ec3d03 seems a better place to introduce it.

Reply:
I'd like to keep it here instead of removing and re-introducing this line.